### PR TITLE
Reduce closures in Math library

### PR DESCRIPTION
**Pull request by gero3**
This Pull Request removes a ton of closures in the math library of three.js. This is no longer needed because modules already enclose the variables. This should improve performance and reduce memory but this should still be tested.

### CODE REVIEW

**Comment by mrdoob** [comment]
Nice!

File: src/math/Box2.js [hunk 1]
```diff
@@ -11,6 +11,8 @@ function Box2( min, max ) {
 
 }
 
+var v1 = new Vector2();
```
Comment (mrdoob) [inline src/math/Box2.js]: Can we add comments in all these?

```js
// Used in setFromCenterAndSize() and distanceToPoint()
var v1 = new Vector2();
```

Comment (gero3) [reply src/math/Box2.js]: Of course

Comment (gero3) [reply src/math/Box2.js]: Added all the comments

**Comment by WestLangley** [comment]
A method that accesses a "global" variable must be careful when calling a method (or a method that calls a method) that also accesses that variable.

You feel comfortable doing this, anyway?

**Comment by gero3** [comment]
@WestLangley I've checked the problem cases for this issue and made use of different variables for these problems

**Comment by donmccurdy** [comment]
> A method that accesses a "global" variable must be careful when calling a method (or a method that calls a method) that also accesses that variable.

Yeah, this makes me nervous... I'm totally willing to believe @gero3 has tested this PR carefully, but I'm less confident that all contributors can collectively avoid shooting ourselves in the foot with future changes.

If we're sure we want to remove these closures (which I think is true, but haven't seen that spelled out explicitly), are there patterns we could borrow from other libraries to make this less of a foot-gun?

For example:

```js
import { ObjectPool } from './ObjectPool.js';
import { Vector2 } from './Vector2.js';

var poolVector2 = new ObjectPool( Vector2 );

class Box3 {

  distanceToPoint ( point ) {

    var v = poolVector2.obtain();

    // ... do stuff ...

    poolVector2.recycle( v );

    return result;

  }

}
```

That's a bit more verbose, but at least there's unambiguous ownership of the pooled objects.

**Comment by mrdoob** [comment]
A problem with the PR is that there are many files being modified at the same time.

I'm comfortable with the `Box2` and `Color` changes, but the other files need more study.

@gero3 any chance you could split this PR in one per file?

**Comment by WestLangley** [comment]
I am not comfortable with this approach for the reasons stated.

What are we gaining by doing this?

**Comment by Mugen87** [comment]
> What are we gaining by doing this?

related #14695

When we want to use ES6 classes at some point in the code base, all such IIFEs have to be removed from the codebase.

**Comment by gero3** [comment]
@mrdoob I can certainly seperate the files in different PRs. I have only sperated Color since that one didn't have any reuse of objects, only some functions.

**Comment by gkjohnson** [comment]
> When we want to use ES6 classes at some point, all such IIFEs have to be removed from the codebase.

@Mugen87 Maybe this belongs in another thread but I've always use this pattern when I've wanted to add IIFE functions to classes:

```js
class foo {}

foo.prototype.bar = (function() {
  const localVariable = {};
  return function() {
    // ...
  };
})();

(new foo()).bar()
```

This should work just as well, right?

**Comment by Mugen87** [comment]
Um, this seems a bit like an anti-pattern to me. When working with classes, there should no reasons to access `prototype`.

**Comment by gkjohnson** [comment]
It's definitely not my favorite either but the syntax for defining classes is more limited than defining the interface via the prototype so it becomes necessary to achieve behavior like this. I've been bit by the shared global cache variables enough that I'm starting to prefer this class IIFE pattern for these cases and if it affords everything we want I'm not sure what the downside is.

It seems that this is a recognized limitation of classes is JS, as well. In the future [class fields](https://github.com/tc39/proposal-class-fields) will alleviate the need for accessing the class prototype and let you define IIFE functions in the class definition itself:

```js
class foo {
    bar = (function() {
        return function() {
            // ...
        };
    })();
}
```

I guess I'm just suggesting that there's a path forward in the future where we IIFEs are a viable option for class functions and if the scoped variables are generally considered a safer and preferred pattern to global cache variables then it may be worth modifying the class prototypes for now.

**Comment by EliasHasle** [comment]
IIFE in classes: https://stackoverflow.com/questions/39297258/iife-in-es6-class-literal/56077521#56077521

**Comment by gkjohnson** [comment]
@EliasHasle I had no idea class properties were supported an any browser, yet - it still looks like that's just available in chrome, though: https://v8.dev/features/class-fields. [This compatibility table](https://kangax.github.io/compat-table/esnext/) shows that Firefox Nightly supports them as well, though it's clearly still cutting edge. I'm not sure what the opinion is on using babel plugins to support cutting edge ES6 features is but there's a plugin to compile class fields available [here](https://babeljs.io/docs/en/babel-plugin-proposal-class-properties).

I think I would consider using the class IIFE pattern to support closure variables with rather than trying to remove all of them from the library, myself.

**Comment by EliasHasle** [comment]
> I think I would consider using the class IIFE pattern to support closure variables with rather than trying to remove all of them from the library, myself.

Yes, that was my thought. Answering the question: "What is the minimal modification needed for migration to classes?". That is if the patterns are in fact equivalent. But @Mugen87 called it an "anti-pattern", putting forth another approach. See https://github.com/mrdoob/three.js/issues/11552#issuecomment-507582135 (consider whether to continue discussion there or here)
I do understand the concerns, which to some extent apply to closures in prototypes too. They do not make the code very readable, and it is easy to forget/misunderstand details on the behavior of the `this` keyword. And the behavior depends on whether one is in strict mode or not. So exploring other solutions is wise.

File: src/math/Box3.js [hunk 2]
```diff
@@ -215,69 +268,15 @@ Object.assign( Box3.prototype, {
...
-						v1.applyMatrix4( node.matrixWorld );
-
-						scope.expandByPoint( v1 );
+	expandByObject: function ( object ) {
 
-					}
+		object.updateMatrixWorld( true );
 
-				} else if ( geometry.isBufferGeometry ) {
+		object.traverse( traverse.bind( this ) );
```
Comment (Mugen87) [inline src/math/Box3.js]: I'm analyzing the closures in `Box3` now. Are you aware that `bind()` returns a new function on each invocation? It's the same like defining `traverse()` directly in `expandByObject()`.

I'm afraid we need some refactoring here...

**Comment by Mugen87** [comment]
Closing in favor of #17159.

**Pull request closed without merging**